src/cpp/r/RExec.cpp

```diff
@@ -126,21 +128,18 @@ enum EvalType {
    EvalTry,    // use R_tryEval
    EvalDirect  // use Rf_eval directly
 };
+
 Error evaluateExpressionsUnsafe(SEXP expr,
                                 SEXP envir,
                                 SEXP* pSEXP,
                                 sexp::Protect* pProtect,
                                 EvalType evalType)
 {
+   ASSERT_MAIN_THREAD;
```

Comment:
If we are not on the main thread, I like that the ASSERT_MAIN_THREAD will include the logError but won't it now just go ahead and execute the expression, rather than returning an rCodeExecutionError?  In my testing, an attempt to actually execute R code off the main thread would longjmp that thread back into the R exec loop and that was not a fun debugging experience!  When we return an error, it tends to just abort the operation, like with that bug saving RMarkdown documents... the only side-effects now are the logError and now I feel like it would crash.

Reply:
Thanks, you're right. I've revised the PR to allow for early returns in these contexts.